browser/brave_wallet/keyring_service_unittest.cc

```diff
@@ -3819,4 +3819,195 @@ TEST_F(KeyringServiceAccountDiscoveryUnitTest, RestoreWalletTwice) {
+  }
+
+  const std::string& saved_mnemonic() const { return saved_mnemonic_; }
+
+ private:
+  std::string saved_mnemonic_;
+  base::test::ScopedFeatureList feature_list_;
+};
+
+TEST_F(KeyringServicePasswordIternationsMigrationUnitTest, NoMigration) {
+  EXPECT_TRUE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+  SetupKeyring();
+  EXPECT_FALSE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+
+  KeyringService service(json_rpc_service(), GetPrefs());
+  EXPECT_TRUE(Unlock(&service, "brave"));
+  ValidateUnlockedKeyring(&service);
+}
+
+TEST_F(KeyringServicePasswordIternationsMigrationUnitTest, MigrateWithUnlock) {
+  // Setup prefs with legacy iterations count value.
+  KeyringService::GetPbkdf2IterationsForTesting() = 100000;
+  SetupKeyring();
+  KeyringService::GetPbkdf2IterationsForTesting() = absl::nullopt;
+
+  // Reset migration pref so migration runs on next unlock.
+  EXPECT_FALSE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+  GetPrefs()->ClearPref(kBraveWalletKeyringsNeedPasswordsMigration);
+  EXPECT_TRUE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+
+  KeyringService service(json_rpc_service(), GetPrefs());
+  EXPECT_TRUE(Unlock(&service, "brave"));
+  EXPECT_FALSE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+
+  service.Lock();
+  // Unlock with wrong password fails.
+  EXPECT_FALSE(Unlock(&service, "another password"));
+  KeyringService::GetPbkdf2IterationsForTesting() = 100000;
+  // Unlock with legacy iterations fails.
+  EXPECT_FALSE(Unlock(&service, "brave"));
+  KeyringService::GetPbkdf2IterationsForTesting() = absl::nullopt;
+
+  // Unlocking again works.
+  EXPECT_TRUE(Unlock(&service, "brave"));
+  EXPECT_FALSE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+
+  ValidateUnlockedKeyring(&service);
+}
+
+TEST_F(KeyringServicePasswordIternationsMigrationUnitTest, MigrateWithRestore) {
+  // Setup prefs with legacy iterations count value.
+  KeyringService::GetPbkdf2IterationsForTesting() = 100000;
+  SetupKeyring();
+  KeyringService::GetPbkdf2IterationsForTesting() = absl::nullopt;
+
+  // Reset migration pref so migration runs on next restore.
+  EXPECT_FALSE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+  GetPrefs()->ClearPref(kBraveWalletKeyringsNeedPasswordsMigration);
+  EXPECT_TRUE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+
+  // Restore wallet.
+  KeyringService service(json_rpc_service(), GetPrefs());
+  EXPECT_TRUE(RestoreWallet(&service, saved_mnemonic(), "brave", false));
+  EXPECT_FALSE(
+      GetPrefs()->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration));
+
+  // Check wallet state after restore.
+  ValidateUnlockedKeyring(&service);
+}
+
+TEST_F(KeyringServicePasswordIternationsMigrationUnitTest,
+       MigrateWithRestoreBadPassword) {
```

Comment:
This is not bad password. If users provide current mnemonic used in the wallet but different password, then we really do resume process with new password.

Reply:
But we are loosing imported accounts then.
I'm still going to rename the test or remove if makes sense after moving migration point

Reply:
Renamed to MigrateWithRestoreAndNewPassword